src/main.cpp

```diff
@@ -2802,6 +2815,24 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
         }
     }
```

Comment:
The [ZIP 221 block header consensus rules](https://zips.z.cash/zip-0221#block-header-semantics-and-consensus-rules) are missing. They need to occur as part of this scope, before the view is updated, because the history tree is overwritten by each block rather than storing one copy per block, and the block needs to commit to the previous block's tree. Something like:
```cpp
if (chainparams.GetConsensus().NetworkUpgradeActive(pindex->nHeight, Consensus::UPGRADE_HEARTWOOD)) {
    auto prevConsensusBranchId = CurrentEpochBranchId(pindex->nHeight - 1, chainparams.GetConsensus());
    auto hashChainHistoryRoot = view.GetHistoryRoot(prevConsensusBranchId);
    if (block.hashLightClientRoot != hashChainHistoryRoot {
        return state.DoS(...)
    }
} else if (chainparams.GetConsensus().NetworkUpgradeActive(pindex->nHeight, Consensus::UPGRADE_SAPLING)) {
    // Existing consensus rule checking against sapling_tree.root()
}
```

Reply:
This will be addressed after I make the chain DB changes.